src/main.rs

```diff
@@ -89,14 +88,30 @@ fn set_flags(args: Vec<String>) -> Vec<String> {
 
 type DenoException<'a> = &'a str;
 
-struct Deno {
-  ptr: *const DenoC,
+pub struct Deno {
+  ptr: *const binding::DenoC,
+  dir: deno_dir::DenoDir,
 }
 
+static DENO_INIT: std::sync::Once = std::sync::ONCE_INIT;
+
 impl Deno {
-  fn new() -> Deno {
-    let ptr = unsafe { deno_new(ptr::null(), deno_handle_msg_from_js) };
-    Deno { ptr: ptr }
+  fn new<'a>() -> &'a mut Deno {
+    DENO_INIT.call_once(|| {
+      unsafe { binding::deno_init() };
+    });
+
+    let deno_box = Box::new(Deno {
+      ptr: 0 as *const binding::DenoC,
+      dir: deno_dir::DenoDir::new(None).unwrap(),
+    });
+    let deno: &'a mut Deno = Box::leak(deno_box);
```

Comment:
Jeez... I can see how this makes for some sleepless nights.
Why is Deno made mutable here if it gets cast to *const one line below?
(Not saying it's wrong, I just don't understand this very well.)

A little comment on what's going on here would be in order too (as in -- mention that C-Deno needs a reference to Rust-Deno. And why this leak is necessary (is it really btw?).

Reply:
It gets returned as mut (but my confidence that this is optimal code is low)